tools/approval-service/internal/service/ttleventcache_test.go

```diff
@@ -0,0 +1,280 @@
+ */
+
+package service
+
+import (
+	"fmt"
+	"sync/atomic"
+	"testing"
+	"testing/synctest"
+	"time"
+)
+
+func (c *TTLEventCache) checkInternalState(t *testing.T, expectedMapLen, expectedListLen int) {
+	t.Helper()
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if len(c.cache) != expectedMapLen {
+		t.Errorf("len(c.cache) = %d; want %d", len(c.cache), expectedMapLen)
+	}
+	if len(c.expiryList) != expectedListLen {
+		t.Errorf("len(c.expiryList) = %d; want %d", len(c.expiryList), expectedListLen)
+	}
+}
+
+func TestTTLEventCache_WithInvalidTLL_ShouldSetDefaultValue(t *testing.T) {
+	defaultTTL := 15 * time.Second
+
+	ecWithZero := NewTTLEventCache(0)
+	if ecWithZero.ttl != defaultTTL {
+		t.Errorf("NewTTLEventCache(0) did not set default TTL: got: %v; want: %v", ecWithZero.ttl, defaultTTL)
+	}
+
+	ecNegative := NewTTLEventCache(-5 * time.Second)
+	if ecNegative.ttl != defaultTTL {
+		t.Errorf("NewTTLEventCache(-5) did not set default TTL: got: %v; want: %v", ecNegative.ttl, defaultTTL)
+	}
+}
+
+func TestTTLEventCache_ValidTTL_ShouldOverrideDefault(t *testing.T) {
+	// Arrange
+	oneSecondTTlForTest := 1 * time.Second
+
+	// Act
+	ecPositive := NewTTLEventCache(oneSecondTTlForTest)
+
+	// Assert
+	if ecPositive.ttl != oneSecondTTlForTest {
+		t.Errorf("NewTTLEventCache(1) did not set correct TTL: got: %v; want: %v", ecPositive.ttl, oneSecondTTlForTest)
+	}
+}
+
+func TestTTLEventCache_TryAdd_HappyPath(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		// Arrange
+		keyForTest := "key-for-test"
+		ec := NewTTLEventCache(15 * time.Second)
+
+		// Act
+		isSuccessfulFirstTryAdd := ec.TryAdd(keyForTest)
+
+		// Assert
+		if !isSuccessfulFirstTryAdd {
+			t.Errorf("TryAdd(key) returned false on first add, expected true")
+		}
+		ec.checkInternalState(t, 1, 1)
+	})
+}
+
+func TestTTLEventCache_TryAdd_BasicRejectDuplicates(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ec := NewTTLEventCache(15 * time.Second)
+		keyForTest := "key-for-test"
+
+		isSuccessfulFirstTryAdd := ec.TryAdd(keyForTest)
+		if !isSuccessfulFirstTryAdd {
+			t.Errorf("TryAdd(key) returned false on first add, expected true")
+		}
+		ec.checkInternalState(t, 1, 1)
+
+		isSuccessfulSecondTryAdd := ec.TryAdd(keyForTest)
+		if isSuccessfulSecondTryAdd {
+			t.Errorf("TryAdd(keyForTest) returned true on duplicate add, expected false")
+		}
+		ec.checkInternalState(t, 1, 1)
+	})
+}
+
+func TestTTLEventCache_TryAdd_Concurrency(t *testing.T) {
+	synctest.Test(t, func(t *testing.T) {
+		ttl := 15 * time.Second
+		ec := NewTTLEventCache(ttl)
+
+		const (
+			distinctKeys   = 5
+			requestsPerKey = 200
+		)
+
+		var successfulAdds int32
+
+		// Schedule one small step per logical request. The synctest harness runs these
+		// steps under a deterministic scheduler so inter-leavings are reproducible.
+		for i := 0; i < distinctKeys; i++ {
+			key := fmt.Sprintf("key-%d", i)
+			for j := 0; j < requestsPerKey; j++ {
+				if ec.TryAdd(key) {
+					atomic.AddInt32(&successfulAdds, 1)
```

Comment:
Is this atomic needed? There doesn't seem to be any concurrency in the test or the unit under test. I'd also rename the "Concurrency" in the test to "Multiple" perhaps since there is no concurrency here.